### PR TITLE
Send screen overhaul

### PR DESCRIPTION
**Pull request by da-kami**
- Decode invoice
- Form validation

Note: we are not displaying the decoded values yet.

TODO:

- [x] Display decoded values once available

---

### CODE REVIEW

File: mobile/lib/features/wallet/application/wallet_service.dart [hunk 1]
```diff
@@ -39,6 +40,17 @@ class WalletService {
     }
   }
 
+  Future<LightningInvoice?> decodeInvoice(String invoice) async {
+    try {
+      LightningInvoice lightningInvoice = await rust.api.decodeInvoice(invoice: invoice);
+      FLog.info(text: "Successfully decoded invoice.");
```
Comment (bonomat) [inline mobile/lib/features/wallet/application/wallet_service.dart:46]: Nit: I guess debug would be enough for this message.

File: mobile/native/src/api.rs [hunk 2]
```diff
@@ -217,3 +224,38 @@ pub fn get_seed_phrase() -> Result<SyncReturn<Vec<String>>> {
...
+
+pub fn decode_invoice(invoice: String) -> Result<LightningInvoice> {
+    anyhow::ensure!(!invoice.is_empty(), "received empty invoice");
+    let invoice = &Invoice::from_str(&invoice).context("Could not parse invoice string")?;
+    let description = match invoice.description() {
+        InvoiceDescription::Direct(direct) => direct.to_string(),
+        InvoiceDescription::Hash(_) => "".to_string(),
+    };
+    // can't use as_millis as the frb does not support u128
+    let timestamp = invoice
```
Comment (bonomat) [inline mobile/native/src/api.rs:244]: What's the timestamp needed for?

**Review by bonomat — APPROVED** [review]
Looks good :)

File: mobile/native/src/api.rs (hunk 2 above)
Comment (da-kami) [reply mobile/native/src/api.rs:244]: I just blindly copied this from the POC - will have another look :)

File: mobile/lib/features/wallet/send_screen.dart [hunk 3]
```diff
@@ -14,33 +21,118 @@ class SendScreen extends StatefulWidget {
...
+
+                    return null;
+                  },
+                ),
+              ),
+              Center(
+                  child: Padding(
+                padding: const EdgeInsets.only(bottom: 10.0, left: 32.0, right: 32.0),
+                child: Text(
+                  "During the beta a minimum of $channelReserve sats have to remain in the wallet."
```
[the remark this answers is not in the document]
Comment (da-kami) [reply mobile/lib/features/wallet/send_screen.dart:104]: I like the suggestions - we should do this for all those popups I suppose, but it's quite some work and I would not give it prio

File: mobile/native/src/api.rs (hunk 2 above)
Comment (da-kami) [reply mobile/native/src/api.rs:244]: In the poc we just displayed it - no other use. 
As far as I understand it's the expiry timestamp of the invoice. 
I would keep it for completion.

Comment (bonomat) [reply mobile/native/src/api.rs:244]: > As far as I understand it's the expiry timestamp of the invoice.

I'm confused :) The expiry is 4 lines below.

Comment (holzeis) [reply mobile/native/src/api.rs:244]: The timestamp is when the invoice has been created and the expiry defines the number of seconds until it is invalid.

The following example should make it clear. 
```
"timestamp": "1496314658",
"expiry": "3600",
```

**Comment by da-kami** [comment]
Note: Fixed up the expiry after taking the gif record :)

**Review by bonomat — APPROVED** [review]
LGTM

**Pull request merged**